--- gendocs_template.orig	2008-06-19 15:48:31.000000000 +0200
+++ gendocs_template	2008-06-19 15:49:05.000000000 +0200
@@ -59,6 +59,14 @@
     (%%TEXI_TGZ_SIZE%%K bytes gzipped tar file).</a></li>
 </ul>
 
+<p>Documentation is also available in a few experimental formats:
+
+<ul>
+  <li><a href="../reference/">HTML API Reference Manual (GTK-DOC)</a>.</li>
+  <li><a href="../reference/gsasl.devhelp">DevHelp API Reference Manual (GTK-DOC)</a>.</li>
+  <li><a href="../components.png">PNG image with developer and API information</a>.</li>
+</ul>
+
 <p>You can <a href="http://shop.fsf.org/">buy printed copies of
 some manuals</a> (among other items) from the Free Software Foundation;
 this helps support FSF activities.</p>
